### PR TITLE
New TimeseriesGenerator

### PR DESCRIPTION
**Pull request by TanguyUrvoy**
**changes relative to previous version of `TimeseriesGenerator()`**
- kept `length` (with a depreciation warning)
- added `hlength` as a replacement
- same parameters order

**Other changes relative to keras version**
- consolidated the `init()` method by adding a host of sanity checks
- now it also works with text
- added prediction `gap` parameter
- shuffle is now a real shuffle (i.e. without replacement) 
- added `target_seq` option for seq2seq models
- added `dtype` parameter to force output dtype
- added `stateful` option to help parameters tuning in stateful mode (experimental)

### CODE REVIEW

File: keras_preprocessing/sequence.py [hunk 1]
```diff
@@ -250,121 +252,245 @@ def _remove_long_seq(maxlen, seq, label):
...
+            raise ValueError('`sampling_rate` must be strictly positive.')
+        if stride <= 0:
+            raise ValueError('`stride` must be strictly positive.')
+        if batch_size <= 0:
+            raise ValueError('`batch_size` must be strictly positive.')
+        if len(data) > len(targets):
+            raise ValueError('`targets` has to be at least as long as `data`.')
+
+        if hlength is None:
+            if length % sampling_rate != 0:
```
Comment (Dref360) [inline keras_preprocessing/sequence.py]: Please add a DeprecationWarning

File: keras_preprocessing/sequence.py [hunk 2]
```diff
@@ -250,121 +252,245 @@ def _remove_long_seq(maxlen, seq, label):
...
+                "Unless you know what you do, `gap` should be zero or"
+                " a multiple of `sampling_rate`.", UserWarning)
+
+        self.hlength = hlength
+        assert self.hlength > 0
+
+        self.data = np.asarray(data)
+        self.targets = np.asarray(targets)
+
+        # FIXME: targets must be 2D for sequences output
```
Comment (Dref360) [inline keras_preprocessing/sequence.py]: Is this a new limitation or was it always like that?

File: tests/sequence_test.py [hunk 3]
```diff
@@ -103,116 +103,158 @@ def test_remove_long_seq():
...
+    with assert_raises(ValueError) as context:
+        TimeseriesGenerator(data, data, length=50, stride=0)
+    error = str(context.exception)
+    print(error)
+    assert 'must be strictly positive.' in error
+
+    with assert_raises(ValueError) as context:
+        TimeseriesGenerator(data, data, length=50, sampling_rate=0)
+    error = str(context.exception)
+    print(error)
```
Comment (Dref360) [inline tests/sequence_test.py]: avoid the printing please

**Comment by Dref360** [comment]
@srjoglekar246, Could I get your input on this PR?

File: keras_preprocessing/sequence.py [hunk 4]
```diff
@@ -250,121 +252,245 @@ def _remove_long_seq(maxlen, seq, label):
...
     stride, length of history, etc., to produce batches for
     training/validation.
-
     # Arguments
         data: Indexable generator (such as list or Numpy array)
             containing consecutive data points (timesteps).
-            The data should be at 2D, and axis 0 is expected
-            to be the time dimension.
+            The data should be convertible into a 1D numpy array,
+            if 2D or more, axis 0 is expected to be the time dimension.
```
Comment (srjoglekar246) [inline keras_preprocessing/sequence.py]: nit: Might make this a _little_ easier to understand if we say "and* axis 0"

File: keras_preprocessing/sequence.py [hunk 5]
```diff
@@ -250,121 +252,245 @@ def _remove_long_seq(maxlen, seq, label):
...
+
+    for i in range(len(data_gen)):
+        print(data_gen[i][0].tostring(), "->'%s'" % data_gen[i][1].tostring())
+
+    assert data_gen[-1][0].shape == (1, 10) and data_gen[-1][1].shape == (1,)
+    assert data_gen[-1][0].tostring() == u" is simple"
+    assert data_gen[-1][1].tostring() == u"."
+
+    t = np.linspace(0,20*np.pi, num=1000) # time
+    x = np.sin(np.cos(3*t)) # input signa
```
Comment (srjoglekar246) [inline keras_preprocessing/sequence.py]: signal*

File: keras_preprocessing/sequence.py [hunk 6]
```diff
@@ -250,121 +252,245 @@ def _remove_long_seq(maxlen, seq, label):
...
+    data_gen = TimeseriesGenerator(txt, txt, hlength=10, batch_size=1, gap=1)
+
+    for i in range(len(data_gen)):
+        print(data_gen[i][0].tostring(), "->'%s'" % data_gen[i][1].tostring())
+
+    assert data_gen[-1][0].shape == (1, 10) and data_gen[-1][1].shape == (1,)
+    assert data_gen[-1][0].tostring() == u" is simple"
+    assert data_gen[-1][1].tostring() == u"."
+
+    t = np.linspace(0,20*np.pi, num=1000) # time
```
Comment (srjoglekar246) [inline keras_preprocessing/sequence.py]: space after ","

File: keras_preprocessing/sequence.py [hunk 7]
```diff
@@ -250,121 +252,245 @@ def _remove_long_seq(maxlen, seq, label):
...
-        self.length = length
+                 batch_size=128,
+                 hlength=None,
+                 target_seq=False,
+                 gap=0,
+                 dtype=None,
+                 stateful=False):
+
+        # Sanity check
+
```
Comment (srjoglekar246) [inline keras_preprocessing/sequence.py]: Remove stray newline

File: tests/sequence_test.py [hunk 8]
```diff
@@ -103,116 +103,158 @@ def test_remove_long_seq():
...
+    assert np.array_equal(data_gen[0][0], np.array(
         [np.array(data[10:19:2]), np.array(data[11:20:2])]))
-    assert (np.allclose(data_gen[0][1],
-                        np.array([targets[20], targets[21]])))
+    assert (np.array_equal(data_gen[0][1],
+                           np.array([targets[20], targets[21]])))
+
+
+def test_TimeseriesGenerator_exceptions():
+
```
Comment (srjoglekar246) [inline tests/sequence_test.py]: Please remove newlines at the beginning of functions :-)

File: tests/sequence_test.py [hunk 9]
```diff
@@ -103,116 +103,158 @@ def test_remove_long_seq():
...
+    x = np.array([i for i in range(7)])
 
-            y = np.concatenate([g[ix][1] for ix in range(len(g))], axis=0)
-            assert_allclose(y, expected)
+    g = TimeseriesGenerator(x, x, hlength=3, batch_size=2)
+
+    expected_len = ceil((len(x) - g.hlength + 1.0) / g.batch_size)
+    print('gap: %i, hlength: %i, expected-len:%i, len: %i' %
+          (g.gap, g.hlength, expected_len, g.len))
+    # for i in range(len(g)):
```
Comment (srjoglekar246) [inline tests/sequence_test.py]: Remove commented code

**Comment by rjmccabe3701** [comment]
Looks like this MR is more far reaching then [this](https://github.com/keras-team/keras-preprocessing/pull/251).  In my MR i simply added a ``inter_batch_stride`` option.

I'm hoping that these changes expose the same behavior I'm after (so the first sample of a batch isn't forced to follow the last sample of the previous batch)?

**Comment by OverLordGoldDragon** [comment]
This class looks like a lost cause in need of a total revision. Correct me if I'm wrong, but these lines will _mix_ the batch dimension with a features dimension (timesteps), which is a big no-no and obliterate training:

```python
import numpy as np

data = np.random.randn(240, 4)  # (timesteps, channels)
length = 6
batch_size = 8
stride = 1
start_index, end_index, index = 6, 40, 0
```
```python
i = start_index + batch_size * stride * index
rows = np.arange(i, min(i + batch_size * stride, end_index + 1), stride)

samples = np.array([data[row - length:row:sampling_rate] for row in rows])
print(samples.shape)
# (8, 6, 4)
```
The only workaround is feeding data compatible with such a manipulation to begin with - as no measurements are ever taken in such a format, it'll require an involved preprocessing, maybe on-the-fly for various `length` settings, which will considerably slow down training. There are other pitfalls I'll spare listing - but as I'm not too familiar with `TimeseriesGenerator`, maybe I'm missing something - feel free to point out.

**Comment by fchollet** [comment]
Closing outdated PR. Note that the Keras Preprocessing repository is no longer in use, instead please just use the dataset utilities provided in `tf.keras.utils` ([see docs](https://keras.io/api/data_loading/)).
